src/dbnode/client/replicated_session.go

```diff
@@ -0,0 +1,301 @@
+		asyncSession := asyncSession // capture var
+		if s.workerPool.GoIfAvailable(func() {
+			err := fn(asyncSession)
+			if err != nil {
+				s.metrics.replicateError.Inc(1)
+				s.log.Error("could not replicate write: %v", zap.Error(err))
+			}
+			s.outCh <- err
+		}) {
+			s.metrics.replicateExecuted.Inc(1)
+		} else {
+			s.metrics.replicateNotExecuted.Inc(1)
+		}
+	}
+	return fn(s.session)
+}
+
+// Write value to the database for an ID
+func (s replicatedSession) Write(namespace, id ident.ID, t time.Time, value float64, unit xtime.Unit, annotation []byte) error {
+	return s.replicate(func(session Session) error {
+		return session.Write(namespace, id, t, value, unit, annotation)
+	})
+}
+
+// WriteTagged value to the database for an ID and given tags.
+func (s replicatedSession) WriteTagged(namespace, id ident.ID, tags ident.TagIterator, t time.Time, value float64, unit xtime.Unit, annotation []byte) error {
+	return s.replicate(func(session Session) error {
+		return session.WriteTagged(namespace, id, tags, t, value, unit, annotation)
+	})
+}
+
+// Fetch values from the database for an ID
+func (s replicatedSession) Fetch(namespace, id ident.ID, startInclusive, endExclusive time.Time) (encoding.SeriesIterator, error) {
+	return s.session.Fetch(namespace, id, startInclusive, endExclusive)
+}
+
+// FetchIDs values from the database for a set of IDs
+func (s replicatedSession) FetchIDs(namespace ident.ID, ids ident.Iterator, startInclusive, endExclusive time.Time) (encoding.SeriesIterators, error) {
+	return s.session.FetchIDs(namespace, ids, startInclusive, endExclusive)
+}
+
+func (s replicatedSession) Aggregate(
+	ns ident.ID, q index.Query, opts index.AggregationOptions,
+) (AggregatedTagsIterator, bool, error) {
+	return s.session.Aggregate(ns, q, opts)
+}
+
+// FetchTagged resolves the provided query to known IDs, and fetches the data for them.
+func (s replicatedSession) FetchTagged(namespace ident.ID, q index.Query, opts index.QueryOptions) (results encoding.SeriesIterators, exhaustive bool, err error) {
+	return s.session.FetchTagged(namespace, q, opts)
+}
+
+// FetchTaggedIDs resolves the provided query to known IDs.
+func (s replicatedSession) FetchTaggedIDs(namespace ident.ID, q index.Query, opts index.QueryOptions) (iter TaggedIDsIterator, exhaustive bool, err error) {
+	return s.session.FetchTaggedIDs(namespace, q, opts)
+}
+
+// ShardID returns the given shard for an ID for callers
+// to easily discern what shard is failing when operations
+// for given IDs begin failing
+func (s replicatedSession) ShardID(id ident.ID) (uint32, error) {
+	return s.session.ShardID(id)
+}
+
+// IteratorPools exposes the internal iterator pools used by the session to clients
+func (s replicatedSession) IteratorPools() (encoding.IteratorPools, error) {
+	return s.session.IteratorPools()
+}
+
+// Close the session
+func (s replicatedSession) Close() error {
+	err := s.session.Close()
+	for _, as := range s.asyncSessions {
+		as.Close()
```

Comment:
Probably want to collect the errors here.  (so you know if closing session fails, shouldn't matter too much on shutdown whether an error is returned in the case of an error with a real async session)